### PR TITLE
Add tests to have full code coverage of the function `_viewbaseset_richcmp`

### PR DESCRIPTION
**Pull request by skinnyBat**
## What do these changes do?

Add tests to improve the code coverage of `_multidict_base_py`, by fully covering the `_viewbaseset_richcmp` function.

## Are there changes in behavior for the user?

None

## Related issue number

Partially addresses https://github.com/aio-libs/multidict/issues/921.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes

### CODE REVIEW

File: multidict/_multidict_base.py [hunk 1]
```diff
@@ -46,6 +46,8 @@ def _viewbaseset_richcmp(view, other, op):
             if elem not in view:
                 return False
         return True
+    else:  # pragma: no cover
+        return NotImplemented
```
Comment (webknjaz) [inline multidict/_multidict_base.py:50]: Why is this needed? Are there any other `op`s that exist?

Should this be `assert_never()` per https://stackoverflow.com/a/74535743/595220?

If there's other `op`s, let's cover this branch by using them in tests.

File: tests/test_multidict.py [hunk 2]
```diff
@@ -281,34 +281,72 @@ def test_keys_is_set_less(self, cls: Type[MutableMultiMapping[str]]) -> None:
...
+    def test_keys_is_set_less_equal(
+        self,
+        cls: Type[MutableMultiMapping[str]],
+        contents: List[Tuple[str, str]],
+        expected: bool,
+    ) -> None:
+        d = cls(contents)
 
-        assert d.keys() <= {"key"}
+        assert (d.keys() <= {"key", "key2"}) == expected
```
Comment (webknjaz) [inline tests/test_multidict.py:301]: For booleans, it is customary to use an identity check, just like with `None`:
```suggestion
        assert (d.keys() <= {"key", "key2"}) is expected
```

Although, I'd probably put the check result into a variable to avoid having complex nested inline code like this.

File: tests/test_multidict.py [hunk 3]
```diff
@@ -281,34 +281,72 @@ def test_keys_is_set_less(self, cls: Type[MutableMultiMapping[str]]) -> None:
...
+        assert (d.keys() >= set_) == expected
+
+    @pytest.mark.parametrize("op", (operator.le, operator.lt, operator.ge, operator.gt))
+    def test_keys_compare_type_error(
+        self, cls: Type[MutableMultiMapping[str]], op: Callable[[object, object], bool]
     ) -> None:
         d = cls([("key", "value1")])
 
-        assert {"key"} >= d.keys()
+        with pytest.raises(TypeError):
```
Comment (webknjaz) [inline tests/test_multidict.py:335]: When using `pytest.raises()`, we should always have a `match=` regex set in the check. With starting `^` and trailing `$`.

File: multidict/_multidict_base.py (hunk 1 above)
Comment (webknjaz) [reply multidict/_multidict_base.py:50]: If this is needed, add a bugfix change note. If not, don't include it and add a contrib change note.

https://github.com/aio-libs/multidict/tree/master/CHANGES#adding-change-notes-with-your-prs

Comment (skinnyBat) [reply multidict/_multidict_base.py]: `op` can only be one of the six values(https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_richcompare). I added this so that the coverage doesn't report that the previous `elif` statement isn't fully covered. The previous `elif` should never to false, as these `if/elif` statements covers all legal values.

`assert_never()` only exists in Python 3.11 and above, so have also added a fallback for versions prior to that. 

As this doesn't change any user behavior, should this be a contrib change note, or a bugfix change note?

Comment (webknjaz) [reply multidict/_multidict_base.py]: Contrib is good. As you mentioned — this doesn't bring end-user facing changes.

File: CHANGES/928.contrib.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Added tests to have full code coverage of the ``_viewbaseset_richcmp`` function -- by :user:`skinnyBat`
```
Comment (webknjaz) [inline CHANGES/928.contrib.rst:1]: We need a small punctuation correction — a trailing period. I also backed in a more specific path to the function:
```suggestion
Added tests to have full code coverage of the
``multidict._multidict_base._viewbaseset_richcmp()`` function
-- by :user:`skinnyBat`.
```

File: multidict/_multidict_base.py [hunk 5]
```diff
@@ -1,5 +1,14 @@
+import sys
 from collections.abc import ItemsView, Iterable, KeysView, Set, ValuesView
 
+if sys.version_info >= (3, 11):
+    from typing import assert_never
+else:
+    from typing import Any, NoReturn
+
+    def assert_never(value: Any) -> NoReturn:
+        raise AssertionError(f"Expected code to be unreachable, but got: {value!r}")
```
Comment (webknjaz) [inline multidict/_multidict_base.py:10]: A typical solution is to use `typing_extensions` to shim this. I'll need to be added to the runtime deps in `setup.cfg` with an environment marker, restricting the Python version.

File: tests/test_multidict.py [hunk 6]
```diff
@@ -281,34 +281,79 @@ def test_keys_is_set_less(self, cls: Type[MutableMultiMapping[str]]) -> None:
...
         d = cls([("key", "value1")])
 
-        assert {"key"} >= d.keys()
+        with pytest.raises(
+            TypeError,
+            match=(
+                "^'.+' not supported between instances of '.*_KeysView' and 'NoneType'$"
+            ),
+        ):
+            op(d.keys(), None)
```
Comment (webknjaz) [inline tests/test_multidict.py:343]: Could you implement a variant of https://github.com/aio-libs/multidict/pull/936#discussion_r1473671524 instead?

This would end up looking cleaner and not needing a `raises()` check. WDYT?

File: multidict/_multidict_base.py (hunk 1 above)
Comment (webknjaz) [reply multidict/_multidict_base.py]: I'm marking this thread as resolved, since I made a fallback suggestion in another place.

File: setup.cfg [hunk 7]
```diff
@@ -43,6 +43,8 @@ classifiers =
 
 [options]
 python_requires = >= 3.7
+install_requires =
+  typing-extensions >= 4.1.0;python_version<'3.11'
```
Comment (webknjaz) [inline setup.cfg:47]:
```suggestion
  typing-extensions >= 4.1.0; python_version < '3.11'
```
